##############
## Author: Hrishikesh A. Lokhande
## Date:11/31/2021
## Description: This program is provided with the book chapter
## titled 'Advanced Bioinformatics Analysis of miRNA-seq data'
## This program runs differential expression analysis on the precursor counts
## produced using the mirDeep2 Galaxy analysis.
##############
library(DESeq2)

Phenotype <-
  read.csv(
    'PhenoType.txt',
    header = TRUE,
    sep = "\t",
    row.names = 1
  ) ### File that has the sample and phenotype map
Phenotype$Group <-
  as.factor(Phenotype$Group) ## Changing to factors (Cancer/Normal)

CountData <-
  read.csv(
    'ExpressionMatrix.csv',
    header = T,
    sep = ",",
    row.names = 1
  ) ## Reading the precursor counts file.

CountData <- CountData[, rownames(Phenotype)]

dds <- DESeqDataSetFromMatrix(countData = CountData,
                              colData = Phenotype,
                              design = ~ Group)   

### dds is an object that holds all the informtion generated by DESeq2 prior to running differential analysis

dds <- DESeq(dds) ## Running the differential expression analysis
results <- results(dds) ## Extract results
results <-
  results[!is.na(results$padj), ] ## Removing all rows with missing adjusted p-value
results <-
  results[order(results$padj), ] ## Sorting based upon the adjusted p-value

write.csv(results, file = 'DE_results.csv') ## Results are writen to file.
